src/native/managed/cdac/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/Loader_1.cs

```
@@ -196,6 +200,76 @@ bool ILoader.TryGetLoadedImageContents(ModuleHandle handle, out TargetPointer ba
         return true;
     }
 
+    private static uint AlignUp(uint value, uint alignment)
+    {
+        if (alignment == 0)
+            throw new ArgumentException("Alignment must be greater than zero.", nameof(alignment));
+        return (value + alignment - 1) & ~(alignment - 1);
+    }
+
+    private static bool IsMapped(Data.PEImageLayout peImageLayout)
+    {
+        return (peImageLayout.Flags & (uint)PEImageFlags.FLAG_MAPPED) != 0;
+    }
+
+    private TargetPointer FindNTHeaders(Data.PEImageLayout imageLayout)
+    {
+        Data.ImageDosHeader dosHeader = _target.ProcessedData.GetOrAdd<Data.ImageDosHeader>(imageLayout.Base);
+        return imageLayout.Base + (uint)dosHeader.Lfanew;
+    }
+
+    private TargetPointer RvaToSection(int rva, Data.PEImageLayout imageLayout)
+    {
+        TargetPointer ntHeadersPtr = FindNTHeaders(imageLayout);
+        Data.ImageNTHeaders ntHeaders = _target.ProcessedData.GetOrAdd<Data.ImageNTHeaders>(ntHeadersPtr);
+        Target.TypeInfo type = _target.GetTypeInfo(DataType.ImageNTHeaders);
+        int offset = type.Fields[nameof(Data.ImageNTHeaders.OptionalHeader)].Offset;
+        TargetPointer section = ntHeadersPtr + (uint)offset + ntHeaders.FileHeader.SizeOfOptionalHeader;
+        TargetPointer sectionEnd = section + _target.GetTypeInfo(DataType.ImageSectionHeader).Size!.Value * ntHeaders.FileHeader.NumberOfSections;
+        uint sectionAlignment = ntHeaders.OptionalHeader.SectionAlignment;
+        while (section < sectionEnd)
+        {
+            Data.ImageSectionHeader sectionHeader = _target.ProcessedData.GetOrAdd<Data.ImageSectionHeader>(section);
+            if (rva < sectionHeader.VirtualAddress + AlignUp(sectionHeader.VirtualSize, sectionAlignment))
```

I see that this is mirrors PEDecoder::RvaToSection, but it does not look right. 

For images that are not mapped, we can map RVA to physical data only if the RVA is between `section->VirtualAddress` and `section->VirtualAddress + section->SizeOfRawData`. If the RVA is outside this range, there is no good mapping. This situation should only ever happen in corrupted images that we do not see in practice. 

As written, this logic is going to produce bogus result that will lead to harder to diagnose crash later. I am wondering whether it is something we want to fix.

---

If that is the case, we can return an error code in the error case.